class.c

```diff
@@ -1059,6 +1059,29 @@ Perl_class_add_ADJUST(pTHX_ HV *stash, CV *cv)
     av_push(aux->xhv_class_adjust_blocks, (SV *)cv);
 }
 
+OP *
+Perl_ck_classname(pTHX_ OP *o)
+{
+    if(!CvIsMETHOD(PL_compcv))
+        croak("Cannot use __CLASS__ outside of a method or field initializer expression");
+
+    return o;
+}
+
+PP(pp_classname)
+{
+    dSP;
+
+    SV *self = PAD_SVl(PADIX_SELF);
+    assert(SvTYPE(SvRV(self)) == SVt_PVOBJ);
+
+    EXTEND(SP, 1);
+    PUSHs(sv_newmortal());
+    sv_ref(*SP, SvRV(self), true);
```

Comment:
Wouldn't it be better to make this use TARG?

Reply:
Ahyes, good call. Have added in a new commit for now; needs squashing before merge.